code/parachain/frame/cosmwasm/src/ibc.rs

```diff
@@ -25,22 +59,325 @@ impl<T: Config> Pallet<T> {
 	}
 
 	pub(crate) fn do_ibc_send_packet(
-		_vm: &mut CosmwasmVM<T>,
-		_channel_id: String,
-		_data: cosmwasm_minimal_std::Binary,
+		vm: &mut CosmwasmVM<T>,
+		channel_id: String,
+		data: cosmwasm_minimal_std::Binary,
 		_timeout: cosmwasm_minimal_std::ibc::IbcTimeout,
 	) -> Result<(), CosmwasmVMError<T>> {
-		Err(Error::<T>::Unsupported.into())
+		let port_id = PortId::from_str(&Self::do_compute_ibc_contract_port(
+			vm.contract_address.as_ref().clone(),
+		))
+		.expect("address is port; qed");
+		let channel_id = ChannelId::from_str(&channel_id)
+			.map_err(|_| CosmwasmVMError::<T>::Ibc("unsupported channel name".to_string()))?;
+
+		T::IbcRelayer::send_packet(SendPacketData {
+			data: data.to_vec(),
+			timeout_timestamp_offset: ((T::UnixTime::now().as_secs() + 36) -
+				T::UnixTime::now().as_secs()) *
+				1_000_000,
+			timeout_height_offset: 0,
+			channel_id,
+			port_id,
+		})
+		.map_err(|_| Error::<T>::Unsupported.into())
 	}
 
 	pub(crate) fn do_ibc_close_channel(
 		_vm: &mut CosmwasmVM<T>,
 		_channel_id: String,
-	) -> Result<(), CosmwasmVMError<T>> {
+	) -> Result<(), CosmwasmVMError<T>> {	
+		/// how to ensure contract closes only own channel? we have contract(port) here, by not map backward	
```

Comment:
We probably need an extra function in `IbcHandler` like `close_channel(channel, port)` as they can recover the channel from the channel   and source port (just like in sendPacket)

Reply:
set dependency on centauri task by david